### PR TITLE
Add support for ed25519 (2/2)

### PR DESCRIPTION
**Pull request by gjgd**
Follow up to https://github.com/decentralized-identity/ion-sdk/pull/20

Adding support for Ed25519 in IonSDK

### CODE REVIEW

File: lib/InputValidator.ts [hunk 1]
```diff
@@ -1,14 +1,29 @@
...
+import JwkEd25519 from './models/JwkEd25519';
 import JwkEs256k from './models/JwkEs256k';
 import OperationKeyType from './enums/OperationKeyType';
 
 /**
  * Class containing input validation methods.
  */
 export default class InputValidator {
+  /**
+   * Validates the schema of a ES256K JWK key.
```
Comment (OR13) [inline lib/InputValidator.ts]:
```suggestion
   * Validates the schema of a Ed25519 or secp256k1 JWK
```

File: lib/InputValidator.ts [hunk 2]
```diff
@@ -1,14 +1,29 @@
...
+   */
+  public static validateOperationKey (operationKeyJwk: JwkEs256k | JwkEd25519, operationKeyType: OperationKeyType) {
+    if (IonKey.isJwkEs256k(operationKeyJwk)) {
+      InputValidator.validateEs256kOperationKey(operationKeyJwk, operationKeyType);
+    } else if (IonKey.isJwkEd25519(operationKeyJwk)) {
+      InputValidator.validateEd25519OperationKey(operationKeyJwk, operationKeyType);
+    } else {
+      throw new IonError(ErrorCode.UnsupportedKeyType, `JWK key should be Es256k or Ed25119.`);
```
Comment (OR13) [inline lib/InputValidator.ts]:
```suggestion
      throw new IonError(ErrorCode.UnsupportedKeyType, `JWK key should be secp256k1 or Ed25519.`);
```

Comment (OR13) [reply lib/InputValidator.ts]: alternatively, refering to the JWS alg instead of the crv.... in that case `ES256K` and `EdDSA`...I find the confusion regarding this in sidetree frustrating...

File: lib/IonDid.ts [hunk 3]
```diff
@@ -15,8 +16,8 @@ export default class IonDid {
    * @param input.document The initial state to be associate with the ION DID to be created using a `replace` document patch action.
    */
   public static createLongFormDid (input: {
-    recoveryKey: JwkEs256k;
-    updateKey: JwkEs256k;
+    recoveryKey: JwkEs256k | JwkEd25519;
```
Comment (OR13) [inline lib/IonDid.ts]: consider moving this or type to a proper type for `SidetreeOperationalPrivateKeyJwk` or similar

File: lib/LocalSigner.ts [hunk 4]
```diff
@@ -12,30 +17,47 @@ export default class LocalSigner implements ISigner {
...
+      const compactJws = await jwsSigner.sign({ data: content });
+      return compactJws;
+    } else if (IonKey.isJwkEd25519(publicKeyJwk)) {
+      const key = await Ed25519KeyPair.from({
+        type: 'JsonWebKey2020',
+        publicKeyJwk,
+        privateKeyJwk: this.privateKey
+      } as any);
+      const signer = key.signer();
+      const jwsSigner = await JWS.createSigner(signer, 'EdDSA', {
```
Comment (OR13) [inline lib/LocalSigner.ts:52]: it might be worth a sanity test with https://github.com/panva/jose in this library to cover JWS interop, we have one here, but you can never test this kind of thing enough...  https://github.com/transmute-industries/verifiable-data/blob/main/packages/jose-ld/src/JWS/jose.test.ts

File: lib/LocalSigner.ts [hunk 5]
```diff
@@ -12,30 +17,47 @@ export default class LocalSigner implements ISigner {
...
+        detached: false,
+        header
+      });
+      const compactJws = await jwsSigner.sign({ data: content });
+      return compactJws;
+    } else {
+      throw new IonError(ErrorCode.UnsupportedKeyType, `JWK key should be Es256k or Ed25119.`);
```
Comment (OR13) [inline lib/LocalSigner.ts]: again, `alg` vs `crv` issue.

File: tests/IonDid.spec.ts [hunk 6]
```diff
@@ -32,6 +35,23 @@ describe('IonDid', async () => {
...
+      const services = [service1];
+
+      const document = {
+        publicKeys: didDocumentKeys,
+        services
+      };
+
+      const longFormDid = IonDid.createLongFormDid({ recoveryKey, updateKey, document });
+
+      const expectedMethodSpecificId = 'did:ion:EiBks49Ah-ZxE1-6Se-PUDk_4_ffylpeqmfMCWUBwwTM-g:eyJkZWx0YSI6eyJwYXRjaGVzIjpbeyJhY3Rpb24iOiJyZXBsYWNlIiwiZG9jdW1lbnQiOnsicHVibGljS2V5cyI6W3siaWQiOiJwdWJsaWNLZXlNb2RlbEVkMjU1MTkiLCJwdWJsaWNLZXlKd2siOnsiY3J2IjoiRWQyNTUxOSIsImt0eSI6Ik9LUCIsIngiOiJoQ3hhTVI2TlVPWEkxQzd3Nlh6MW1jaGMyd1M4RlZ2WGhuZ3NnWjBodHNZIn0sInB1cnBvc2VzIjpbImF1dGhlbnRpY2F0aW9uIiwia2V5QWdyZWVtZW50Il0sInR5cGUiOiJKc29uV2ViS2V5MjAyMCJ9XSwic2VydmljZXMiOlt7ImlkIjoic2VydmljZTFJZCIsInNlcnZpY2VFbmRwb2ludCI6Imh0dHA6Ly93d3cuc2VydmljZTEuY29tIiwidHlwZSI6InNlcnZpY2UxVHlwZSJ9XX19XSwidXBkYXRlQ29tbWl0bWVudCI6IkVpQ0VEZkdaZkdxeXJuMmVsTU1MbGRfMGxQVE8xTXlQbi1MdFlhZWJEYl9xQncifSwic3VmZml4RGF0YSI6eyJkZWx0YUhhc2giOiJFaUI2TVJSNFc5NFhhc21obVhXM2NhUlluWDdUWGYyZktwcUhuZGhqTmcwb2R3IiwicmVjb3ZlcnlDb21taXRtZW50IjoiRWlEbHNIVkFScDgwYTF2azhHQTlxMWcwaFNGbEp3VGdlQkZOMmkyME9sMlRJUSJ9fQ';
```
Comment (OR13) [inline tests/IonDid.spec.ts:51]: the first Long Form ION DID to support EdDSA / ED25519 ?!?!?!?!?! :)

**Review by OR13 — CHANGES_REQUESTED** [review]
Main issue is with `alg` vs `crv` in the variable names... @thehenrytsai how should this be solved? normalize to `alg` ? in other others `EdDSA` everywhere you see `Ed25519` ?

File: lib/IonDid.ts (hunk 3 above)
Comment (gjgd) [reply lib/IonDid.ts]: Added SidetreeKeyJwk type https://github.com/decentralized-identity/ion-sdk/pull/23/commits/bd6625238dc73c9f6ac5962cd3b9d22a87da588a

File: lib/LocalSigner.ts (hunk 5 above)
Comment (gjgd) [reply lib/LocalSigner.ts]: Updated https://github.com/decentralized-identity/ion-sdk/pull/23/commits/9825c9bbd6c156a777d5082f16b52e13cba87dc7

**Comment by gjgd** [comment]
@OR13 I addressed your comments. I am not so sure how to deal with the crv vs alg problem because Ed25519 is the name of the digital signature algorithm, so I think it does belong next to ES256k.

**Comment by OR13** [comment]
@gjgd https://www.iana.org/assignments/jose/jose.xhtml

@thehenrytsai @csuwildcat thoughts: 

```
EdDSA | EdDSA signature algorithms | alg
-- | -- | --
```

```
Ed25519 | Ed25519 signature algorithm key pairs | Optional
-- | -- | --
```

```
ES256K | ECDSA using secp256k1 curve and SHA-256
-- | --
```

```
secp256k1 | SECG secp256k1 curve | Optional
-- | -- | --
```

**Comment by gjgd** [comment]
Screenshot 2022-01-24 at 13 55 34

This is unfortunate naming. Pretty sure Ed25519 is the name of the digital signature algorithm and it should have been `Curve25519` as the name of the associated curve.

ES256k is the implementation of the ECDSA signature algorithm that uses the secp256k1 curve
Ed25519 is the implementation of the EdDSA signature algorithm that uses the Curve25519 curve

Edit: from https://en.wikipedia.org/wiki/EdDSA
> Ed25519 is the EdDSA signature scheme using SHA-512 (SHA-2) and Curve25519

**Comment by OR13** [comment]
@gjgd you are correct, but ION only uses JWS algs... not raw signatures... hence the correct term in ION for a signature is a JWS alg... if ION supported raw ECDSA or EdDSA without a JOSE header this would be a different story.

**Review by csuwildcat — APPROVED** [review]

**Comment by thehenrytsai** [comment]
Thanks @gjgd and @OR13. I am now aware of the PR, will try take a look tomorrow.

**Comment by thehenrytsai** [comment]
Forgive me if the question is silly, but shouldn't the reference implementation be updated first to support ED25519 before SDK claims to have the support? Else anyone using the SDK might be stuck with their DID document?

**Comment by gjgd** [comment]
Planning to use the ion sdk in a project that's not the reference implementation. I assume the reference implementation would throw an error for keys that are not supported?

**Comment by csuwildcat** [comment]
> Planning to use the ion sdk in a project that's not the reference implementation. I assume the reference implementation would throw an error for keys that are not supported?

If you mean the actual ION node code, yes, that will throw if it has not been updated and a release is not propagated across the community in a network upgrade.

**Comment by gjgd** [comment]
I can add a `console.warning` indicating that an ion node may reject an operation with ed25519 keys until the reference implementation supports it, would that work?

**Comment by csuwildcat** [comment]
> I can add a `console.warning` indicating that an ion node may reject an operation with ed25519 keys until the reference implementation supports it, would that work?

There's certainly nothing wrong with Sidetree in general supporting Ed25519, and I'd even like to see ION support it, but my concern is that if we allow people to create long-form IDs in a way that isn't actually valid at the network/protocol level, it will cause strange issues. The best thing would be to see the node code updated with matching support so it actually does work, then push them out together. What's the plan on the node implementation side?

**Comment by gjgd** [comment]
I understand the concerns and the potential confusion this could create, but are you suggesting that we cannot merge this PR (that is already approved) until Sidetree nodes support Ed25519?

If not, please suggest a set of changes that would alleviate this concerns. My suggestion was to log a warning so that developers would know not to use ion-sdk's Ed25519 methods until the nodes support it.

If yes, why weren't this concerns brought up 2+ months ago [when I started to work on this](https://github.com/decentralized-identity/ion-sdk/pull/20)?

**Comment by thehenrytsai** [comment]
@gjgd supporting ED25519 would be a great feature add to ION, so thank you for working on this, while I don't understand (still don't) why you are doing all this work for free, but wasn't going to complain! So we'd love to have the code merged in as soon as possible, but in principle not before actual ION node support. Having an SDK that allows people to create operations that are going to be rejected by ION just feels wrong. I don't understand why wouldn't we have support for ED25519 in ION itself before adding support for it in ION-SDK, maybe you can elaborate why you think it should be done in reverse? 

> If yes, why weren't this concerns brought up 2+ months ago https://github.com/decentralized-identity/ion-sdk/pull/20?

My apologies for not making expectations clear in my comment in the previous PR: "The PR looks great overall, given that it really isn't hooked up with anything really, so I see this more as a preparation to more PRs to come".

Basically the PR was approved on the basis that the code was not hooked up to any public facing paths (in fact I thought what you did was clever due to that fact), and by "more PRs" I was imagining both in Sidetree repo as well as ION-SDK repo, and IMO order is important here.

**Comment by csuwildcat** [comment]
This thread is confusing to me - I am fully supportive of Ed25519 being added as an operation signing option, but was there a plan to have a PR to make it so it was actually useful to folks? I ask because without a matching PR for the node side of things, it seems like we're handing people something confusing that wouldn't actually work as intended. Can't you provide some clarity on your thinking @gjgd?

**Comment by OR13** [comment]
We'll keep an eye on this repo and wait till you implement support for Ed25519, until then we will use the fork where the feature is available.

We use this library / the fork for 2 other did methods that are not ION, one on Ethereum, and one on Amazon QLDB.

We also don't use the ION codebase for either of those since it bundles bitcoin dependencies.

Its might have been more accurate to call this package a `sidetree-did-method-sdk` since that is how we use it... but I can see the logic in not accepting a PR here when the ION side is not supported.

I can't speak for @gjgd , but this is an open source repo and I think it's fine to close a PR when it's not making progress... or when you feel like the suggested changes are more than you have time to volunteer to do for free.

**Comment by gjgd** [comment]
Was planning to use this PR to generate fixtures, that will be used to test a node implementation. I do think this change can stand on its own, but apparently we disagree on that point. 

I hear the concern that the change could be confusing until Ed25519 is implemented on the node side. My suggestion to address this is to:
- Make generating secp256K1 keys a default in [IonKey](https://github.com/decentralized-identity/ion-sdk/blob/0af90f7944fa45a9151d782dd70455d9481ed71f/lib/IonKey.ts#L13) so that developers don't accidentally generate Ed25519 keys for signing Sidetree operations.
- Emit a warning when generating Ed25519 keys until the node implementation is complete.

If this is acceptable, happy to reopen this PR. But to be clear, I'm no longer willing to commit to more work Ed25519 support on the DIF sidetree repo, for the reasons I mentioned on Slack.

**Comment by csuwildcat** [comment]
@thehenrytsai @gjgd what if we accepted the PR with the mod to warn, defaulted to secp, and basically buried the alternate key path until we have ION node support without any docs that say it exists? Would this be an acceptable compromise?
